R/build_copy_queries.R

```diff
@@ -0,0 +1,143 @@
+          ")")) %>%
+      group_by(table) %>%
+      summarize(unique_defs = paste(unique_def, collapse = ",\n  "))
+
+    # foreign key definitions and indexing queries
+    if (is_duckdb(con)) {
+      if (nrow(fks)) {
+        warn("duckdb doesn't support foreign keys, these won't be set in the remote database but are preserved in the `dm`")
+      }
+      # setup ulterior left join so it'll create a NA col for `fk_def`
+      fk_defs <- tibble(table = character(0), fk_defs = character(0))
+      index_queries <- tibble(
+        table = character(0),
+        remote_table = character(0),
+        remote_table_unquoted = character(0),
+        index_name = character(0),
+        sql = character(0))
+    } else {
+      fk_defs <-
+        fks %>%
+        transmute(
+          table = child_table,
+          fk_def = paste0(
+            "FOREIGN KEY (",
+            quote_enum_col(child_fk_cols),
+            ") REFERENCES ",
+            unlist(table_names[parent_table]),
+            " (",
+            quote_enum_col(parent_key_cols),
+            ")")) %>%
+        group_by(table) %>%
+        summarize(fk_defs = paste(fk_def, collapse = ",\n  "))
+
+      index_queries <-
+        fks %>%
+        mutate(
+          index_name = map_chr(child_fk_cols, paste, collapse = "_")) %>%
+        transmute(
+          table = child_table,
+          remote_table = unlist(table_names[table]) %||% character(0),
+          remote_table_unquoted = map_chr(DBI::dbUnquoteIdentifier(con, DBI::SQL(remote_table)), ~ .x@name[["table"]]),
+          index_name = make.unique(paste0(remote_table_unquoted, "__", index_name), sep = "__"),
+          sql = DBI::SQL(paste0(
+            "CREATE INDEX ",
+            index_name,
+            " ON ",
+            remote_table,
+            " (",
+            quote_enum_col(child_fk_cols),
+            ")"
+          ))
+        )
+    }
+  }
+  ## compile `CREATE TABLE ...` queries
+  create_table_queries <-
+    col_defs %>%
+    left_join(pk_defs, by = "table") %>%
+    left_join(unique_defs, by = "table") %>%
+    left_join(fk_defs, by = "table") %>%
+    group_by(table) %>%
+    mutate(
+      remote_table = table_names[table],
+      all_defs = paste(
+        Filter(
+          Negate(is.na),
+          c(col_defs, pk_defs, unique_defs, fk_defs)),
+        collapse = ",\n  ")) %>%
+    ungroup() %>%
+    transmute(table, remote_table, sql = DBI::SQL(glue(
+      "CREATE {if (temporary) 'TEMP ' else ''}TABLE {unlist(remote_table)} (\n  {all_defs}\n)")))
+
+  ## Reorder queries according to topological sort so pks are created before associated fks
+  graph <- create_graph_from_dm(dm, directed = TRUE)
+  topo <- igraph::topo_sort(graph, mode = "in")
+  idx <- match(names(topo), create_table_queries$table)
+
+  if (length(idx) == nrow(create_table_queries)) {
+    create_table_queries <- create_table_queries[idx, ]
+  }
+
+  ## Return a list of both type of queries
+  lst(create_table_queries, index_queries)
```

Comment:
Can we return a tibble here, with one row per table, and perhaps columns `name`, `columns` (list of character), `sql_table` (character), `sql_index` (character)?

Reply:
done, it contains a couple additional column, `remote_name` is needed for populating tables and `index_name` to check for duplicate indexes. `sql_index` is not a character but a character list since a table can have several index queries in general.